### PR TITLE
[State Restoration] Restorable FormField and TextFormField

### PR DESCRIPTION
**Pull request by shihaohong**
Makes `FormField`, and `TextFormField` state restorable. An interesting situation with this is that `FormField` contains an internal generically typed `_value`. I'm not sure how to handle in a straightforward way since there's no guarantee that `T` is serializable. So to deal with this in `TextFormField`, I explicitly call `setValue` in `restoreState` to make sure that the internal value in `FormFieldState` is kept in sync with the `_controller` value whenever `restoreState` is called.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/text_form_field.dart [hunk 1]
```diff
@@ -309,21 +315,27 @@ class TextFormField extends FormField<String> {
 }
 
 class _TextFormFieldState extends FormFieldState<String> {
-  TextEditingController? _controller;
+  late final RestorableTextEditingController _controller = RestorableTextEditingController(text: widget.initialValue);
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/text_form_field.dart:318]: We should only have this instantiated if we actually need it, similarly to how we do it in the Textfield itself: https://github.com/flutter/flutter/pull/63401/files

File: packages/flutter/lib/src/widgets/form.dart [hunk 2]
```diff
@@ -68,7 +70,7 @@ import 'will_pop_scope.dart';
 ///
 ///  * [GlobalKey], a key that is unique across the entire app.
 ///  * [FormField], a single form field widget that maintains the current state.
-///  * [TextFormField], a convenience widget that wraps a [TextField] widget in a [FormField].
+///  * `TextFormField`, a convenience widget that wraps a [TextField] widget in a [FormField].
```
[the remark this answers is not in the document]
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/form.dart:73]: Oops, I think this change was introduced by accident. I'll revert it

**Comment by shihaohong** [comment]
Should be ready for another round of review!

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by shihaohong** [comment]
The `Google testing` failures appear to be flakes. I reran them ([internal link](https://test.corp.google.com/ui#id=OCL:368501636:BASE:368533532:1618443673439:3a4ef0b2)) and they're passing.

**Pull request merged**